python/tests/io/test_pyarrow.py

```diff
@@ -1330,3 +1341,485 @@ def test_pyarrow_wrap_fsspec(example_task: FileScanTask, table_schema_simple: Sc
+    assert datafile.upper_bounds[1].decode() == "zzzzzzzzzzzzzzzzzzzz"
+    assert datafile.upper_bounds[2] == STRUCT_DOUBLE.pack(100)
+
+
+def test_metrics_mode_non_default_trunc(example_table_metadata_v2: Dict[str, Any]) -> None:
+    (file_bytes, metadata, mapping) = construct_test_table()
+
+    datafile = DataFile()
+    table_metadata = TableMetadataUtil.parse_obj(example_table_metadata_v2)
+    table_metadata.properties["write.metadata.metrics.default"] = "truncate(2)"
+    fill_parquet_file_metadata(
+        datafile,
+        metadata,
+        mapping,
+        len(file_bytes),
+        table_metadata,
+    )
+
+    assert len(datafile.value_counts) == 5
+    assert len(datafile.null_value_counts) == 5
+    assert len(datafile.nan_value_counts) == 0
+
+    assert len(datafile.lower_bounds) == 2
+    assert datafile.lower_bounds[1].decode() == "aa"
+    assert datafile.lower_bounds[2] == STRUCT_DOUBLE.pack(1.69)[:2]
+
+    assert len(datafile.upper_bounds) == 2
+    assert datafile.upper_bounds[1].decode() == "zz"
+    assert datafile.upper_bounds[2] == STRUCT_DOUBLE.pack(100)[:2]
+
+
+def test_column_metrics_mode(example_table_metadata_v2: Dict[str, Any]) -> None:
+    (file_bytes, metadata, mapping) = construct_test_table()
+
+    datafile = DataFile()
+    table_metadata = TableMetadataUtil.parse_obj(example_table_metadata_v2)
+    table_metadata.properties["write.metadata.metrics.default"] = "truncate(2)"
+    table_metadata.properties["write.metadata.metrics.column.strings"] = "none"
+    fill_parquet_file_metadata(
+        datafile,
+        metadata,
+        mapping,
+        len(file_bytes),
+        table_metadata,
+    )
+
+    assert len(datafile.value_counts) == 4
+    assert len(datafile.null_value_counts) == 4
+    assert len(datafile.nan_value_counts) == 0
+
+    assert len(datafile.lower_bounds) == 1
+    assert datafile.lower_bounds[2] == STRUCT_DOUBLE.pack(1.69)[:2]
+
+    assert len(datafile.upper_bounds) == 1
+    assert datafile.upper_bounds[2] == STRUCT_DOUBLE.pack(100)[:2]
+
+
+def test_offsets() -> None:
+    (file_bytes, metadata, mapping) = construct_test_table()
+
+    datafile = DataFile()
+    fill_parquet_file_metadata(datafile, metadata, mapping, len(file_bytes))
+
+    assert datafile.split_offsets is not None
+    assert len(datafile.split_offsets) == 1
+    assert datafile.split_offsets[0] == 4
+
+
+def test_dataset() -> pa.Buffer:
+    schema = pa.schema([pa.field("ints", pa.int64()), pa.field("even", pa.bool_())])
+
+    _ints = [0, 2, 4, 8, 1, 3, 5, 7]
+    parity = [True, True, True, True, False, False, False, False]
+
+    table = pa.Table.from_pydict({"ints": _ints, "even": parity}, schema=schema)
+
+    visited_paths = []
+
+    def file_visitor(written_file: Any) -> None:
+        visited_paths.append(written_file)
+
+    with TemporaryDirectory() as tmpdir:
+        pq.write_to_dataset(table, tmpdir, partition_cols=["even"], file_visitor=file_visitor)
```

Comment:
Let's leave out partitioning for now. We first have to investigate how to handle applying the partition transforms.